### PR TITLE
Allow group traversal from training tasks dependencies during Taskcluster task group publication

### LINKED ISSUE #312: Publish past experiment directly from Taskcluster
Depends #309 

We need a new publication mode that relies only on a Taskcluster task group ID to find all relevant tasks and their artifacts (configurations, logs and metrics) and publish them on Weight & Biases.

This will not run inside Taskcluster, but help to publish existing experiments into Weight & Biases without relying on model/experiment/logs directory.

Past experiments to publish:
- [en-hu](https://firefox-ci-tc.services.mozilla.com/tasks/groups/aY25-4fXTcuJNuMcWXUYtQ)
- [lt-en 1](https://firefox-ci-tc.services.mozilla.com/tasks/groups/EW7qV3U5SBSjegnTxGZHkw)
- [lt-en 2](https://firefox-ci-tc.services.mozilla.com/tasks/groups/V-OmRM1yS_GwESTVDPl0VQ)
- [lt-en 3](https://firefox-ci-tc.services.mozilla.com/tasks/groups/K1iHndFUSxSEDRLg_H9l1A)
- [lt-en 4](https://firefox-ci-tc.services.mozilla.com/tasks/groups/DXbS0zreSGSVYloAF8gwJg)

### LINKED ISSUE #448: Traversal tasks from Taskcluster task group publication
Discussed on #312, the idea is to traverse task groups related to the main task group and publish everything.

Single Taskcluster task group publication has been done in #406, you can work from branch `vrigal:taskcluster-publication` directly.

The default behavior of `translations_parser.cli.taskcluster_group` must follow https://github.com/mozilla/firefox-translations-training/issues/312#issuecomment-1919799719. This should be the default behavior, an argument `disable_traversal_groups` should allow to disable it.

### PR DESCRIPTION
**Pull request by EvaBardou**
Closes #448 

I've added a `--no-recursive-lookup` command option to @vrigal's implementation from #406. 

I've produced results with and without group traversal using the same top level task group ID `aY25-4fXTcuJNuMcWXUYtQ`:
-  `parse_tc_group aY25-4fXTcuJNuMcWXUYtQ` => **WandB** results are available [here](https://wandb.ai/teklia/en-hu-PR-461-recursive-lookup-enabled).
- `parse_tc_group aY25-4fXTcuJNuMcWXUYtQ --no-recursive-lookup` => **WandB** results are available [here](https://wandb.ai/teklia/en-hu-PR-461-recursive-lookup-disabled).

### CODE REVIEW

File: tracking/translations_parser/cli/taskcluster_group.py [hunk 1]
```diff
@@ -95,109 +104,186 @@ def get_metrics_from_task(task: dict) -> list[Metric]:
...
-    config = task_group["extra"]["action"]["context"]["input"]
-    experiment = config["experiment"]
-    project = f"{experiment['src']}-{experiment['trg']}"
-    group_name = f"{experiment['name']}_{args.group_id}"
 
-    logger.info(f"Listing completed tasks from group {args.group_id}")
-    resp = queue.listTaskGroup(args.group_id)
-    tasks = resp["tasks"]
-    continuation_token = resp.get("continuationToken")
+def get_training_tasks(group_id: str, grouped_tasks: dict[str, list[dict]]) -> list[list[dict]]:
```
Comment (vrigal) [inline tracking/translations_parser/cli/taskcluster_group.py:137]: I would rename this to `list_training_tasks`. Same for above function

File: tracking/translations_parser/cli/taskcluster_group.py [hunk 2]
```diff
@@ -95,109 +104,186 @@ def get_metrics_from_task(task: dict) -> list[Metric]:
...
+
+    # Browse task dependencies
+    for dependent_task_id in task['dependencies']:
+        dependent_status = queue.status(dependent_task_id)
+
+        group_id = dependent_status["status"]["taskGroupId"]
+        if group_id in known:
+            continue
+
+        known.add(group_id)
```
Comment (vrigal) [inline tracking/translations_parser/cli/taskcluster_group.py:265]: I think it is more explicit to yield each group, and iterate at above level adding elements to the set. \
Passing objects by references seems hard to follow reading code IMO.

**Review by vrigal — COMMENTED** [review]
Great. I saw many variable have been renamed, which makes changes a bit harder to track. Those namings are much better otherwise, thanks.

File: tracking/translations_parser/cli/taskcluster_group.py (hunk 1 above)
Comment (EvaBardou) [reply tracking/translations_parser/cli/taskcluster_group.py]: Fixed in 0df72095f51f41de3e8d9434797a66dea560289f

File: tracking/translations_parser/cli/taskcluster_group.py (hunk 2 above)
Comment (EvaBardou) [reply tracking/translations_parser/cli/taskcluster_group.py:271]: Fixed in 0df72095f51f41de3e8d9434797a66dea560289f

File: tracking/translations_parser/cli/taskcluster_group.py [hunk 3]
```diff
@@ -95,109 +104,200 @@ def get_metrics_from_task(task: dict) -> list[Metric]:
...
+    # Browse task dependencies
+    for dependent_task_id in task["dependencies"]:
+        dependent_status = queue.status(dependent_task_id)
+
+        group_id = dependent_status["status"]["taskGroupId"]
+        if group_id in known:
+            continue
+
+        yield group_id
+        known.add(group_id)
```
Comment (vrigal) [inline tracking/translations_parser/cli/taskcluster_group.py:271]:
```suggestion
```

**Review by vrigal — APPROVED** [review]

File: tracking/translations_parser/cli/taskcluster_group.py [hunk 4]
```diff
@@ -95,109 +104,200 @@ def get_metrics_from_task(task: dict) -> list[Metric]:
...
+
+        yield from list_dependent_group_ids(dependent_task_id, known)
```
[the remark this answers is not in the document]
Comment (EvaBardou) [reply tracking/translations_parser/cli/taskcluster_group.py]: I've tried out your suggestion, but this considerably increases the execution time of the command.

**Before adding the suggestion:**
```
parse_tc_group aY25-4fXTcuJNuMcWXUYtQ  4,67s user 1,29s system 26% cpu 22,851 total
```

**After:**
```
parse_tc_group aY25-4fXTcuJNuMcWXUYtQ  25,43s user 1,56s system 19% cpu 2:15,33 total
```

This is because we don't share the knowledge (discovered groups), in real time, across recursion branches. As discussed, I've only added a comment (see e627864d635b5e9399a937bd0a75dbcc4b79ab60) to clarify this behavior and kept the original implementation as-is.

**Review by vrigal — APPROVED** [review]

**Comment by eu9ene** [comment]
I think we should skip uploading if the group is empty for now (meaning it doesn't include any W&B runs except the `group_logs`). It's not very useful to have them even though they include the config in the group logs but makes it harder to navigate the experiments.

Screenshot 2024-02-27 at 11 59 53 AM

**Review by eu9ene — CHANGES_REQUESTED** [review]
It looks good overall, but there are some things we should fix before merging.

In addition to the comment about removing the empty groups, please update the unit tests to reflect the changes in this PR.

**Comment by EvaBardou** [comment]
Hi @eu9ene, thanks a lot for the review!

> It looks good overall, but there are some things we should fix before merging.
> 
> In addition to the comment about removing the empty groups, please update the unit tests to reflect the changes in this PR.

If I'm not mistaken, it seems that this command is not being tested at all. @vrigal mentioned that it would be complicated to mock the Taskcluster queue in #406 description:
>  I wasn't able to add a test. Mocking Taskcluster queue was complex and I prefer work on other important topics.

What should I do then?

**Comment by EvaBardou** [comment]
> I think we should skip uploading if the group is empty for now (meaning it doesn't include any W&B runs except the `group_logs`). It's not very useful to have them even though they include the config in the group logs but makes it harder to navigate the experiments.

Fixed in 015587036da7906636e9be2d0590d4bbf359b1f6

New results for the `parse_tc_group aY25-4fXTcuJNuMcWXUYtQ` command are available [here](https://wandb.ai/teklia/en-hu-PR-461-skip-empty-groups).

**Comment by eu9ene** [comment]
> Hi @eu9ene, thanks a lot for the review!
> 
> > It looks good overall, but there are some things we should fix before merging.
> > In addition to the comment about removing the empty groups, please update the unit tests to reflect the changes in this PR.
> 
> If I'm not mistaken, it seems that this command is not being tested at all. @vrigal mentioned that it would be complicated to mock the Taskcluster queue in #406 description:
> 
> > I wasn't able to add a test. Mocking Taskcluster queue was complex and I prefer work on other important topics.
> 
> What should I do then?

Ok, let's skip testing for now. The main scenario will be online tracking anyway.

**Review by eu9ene — APPROVED** [review]
Let's update with the main branch and it's good to go.

**Comment by EvaBardou** [comment]
> Let's update with the main branch and it's good to go.

Done
